review/get_review_ids.py

```diff
@@ -0,0 +1,235 @@
+    for record in StudentModule.objects.filter(**problem_filter):
+        state = json.loads(record.state)
+        # The key 'selected' shows up if a problem comes from a
+        # library content module. These cause issues so we skip this.
+        # Issue: Library content contains problems but the CSM brings up
+        # the library content and not the problems within
+        if 'selected' not in state:
+            yield record.module_state_key, state
+
+
+def enroll_user_in_review_course(user, current_course):
+    '''
+    If the user is not enrolled in the review version of the course,
+    they are unable to see any of the problems. This ensures they
+    are enrolled so they can see review problems.
+
+    Parameters:
+        user (User): the current user interacting with the review XBlock
+        current_course (CourseLocator): The course the learner is currently in
+    '''
+    enrollment_course_id = ENROLLMENT_COURSE_MAPPING[str(current_course)]
+    enrollment_status = get_enrollment(user.username, enrollment_course_id)
+    if not enrollment_status:
+        add_enrollment(user.username, enrollment_course_id)
+    elif not enrollment_status['is_active']:
+        update_enrollment(user.username, enrollment_course_id, is_active=True)
+
+
+def delete_state_of_review_problem(user, current_course, problem_id):
+    '''
+    Deletes the state of a review problem so it can be used infinitely
+    many times.
+
+    Parameters:
+        user (User): the current user interacting with the review XBlock
+        current_course (CourseLocator): The course the learner is currently in
+        problem_id (str): The problem id whose state should be cleared
+    '''
+    review_course = current_course.replace(course=current_course.course+'_review')
+    review_key = review_course.make_usage_key('problem', problem_id)
+    try:
+        module_to_delete = StudentModule.objects.get(
+            student_id=user.id,
+            course_id=review_course,
+            module_state_key=review_key
+        )
+        module_to_delete.delete()
+    except StudentModule.DoesNotExist:
+        # The record will not exist in the StudentModule if the learner has not
+        # seen it as a review problem yet so we just want to skip since there
+        # is no state to delete
+        pass
+
+
+def get_correctness_and_attempts(state):
+    '''
+    From the state of a problem from the Courseware Student Module,
+    determine if the learner correctly answered it initially and
+    the number of attempts they had for the original problem
+
+    Parameter:
+        state (dict): The state of a problem
+
+    Returns a tuple of (correct, attempts)
+        correct (Bool): True if correct, else False
+        attempts (int): 0 if never attempted, else number of times attempted
+    '''
+    correct = (state['score']['raw_earned'] == state['score']['raw_possible'])
```

Comment:
Have you found this to be more accurate than checking `grade` and `max_grade` values on the CSM record?
Have you found cases where `raw_possible` is actually 0 on actual problems?

Reply:
If I remember correctly, the reason I went with `raw_earned` and `raw_possible` was that `grade` only shows up after a learner has answered a problem and I want problems a learner never answered to also be eligible and `max_grade` also only shows up if the course team specifically defined it. `raw_earned` and `raw_possible` were always present.
I have not found any cases where `raw_possible` is 0 as of now. And even if it is, I imagine that would mean it is ungraded in which case it would be valid to show anyway.